### PR TITLE
Search term handling implementation

### PR DESCRIPTION
**Pull request by swnsma**
Search term implementation and refactoring

Original PR: https://github.com/magento/inventory/pull/2756

### CODE REVIEW

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CityHandler.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+use Magento\Framework\DataObject;
+use Magento\InventoryInStorePickupApi\Model\SearchResult\Area\SearchTerm\HandlerInterface;
+
+/**
+ * Extract city from Search Term.
+ */
+class CityHandler implements HandlerInterface
+{
+    public const CITY = 'city';
```
Comment (sidolov) [inline InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CityHandler.php]: Please, make const private

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CountryHandler.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+use Magento\Directory\Helper\Data;
+use Magento\Framework\DataObject;
+use Magento\InventoryInStorePickupApi\Model\SearchResult\Area\SearchTerm\HandlerInterface;
+
+/**
+ * Provide Default country for area search.
+ */
+class CountryHandler implements HandlerInterface
+{
+    public const COUNTRY = 'country';
```
Comment (sidolov) [inline InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CountryHandler.php]: Please, make const private

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/PostCodeHandler.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+use Magento\Directory\Model\Country\Postcode\ValidatorInterface;
+use Magento\Framework\DataObject;
+use Magento\InventoryInStorePickupApi\Model\SearchResult\Area\SearchTerm\HandlerInterface;
+
+/**
+ * Extract postcode from search term if search term match postcode validation for store country.
+ */
+class PostCodeHandler implements HandlerInterface
+{
+    public const POSTCODE = 'postcode';
```
Comment (sidolov) [inline InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/PostCodeHandler.php]: Please, make const private

File: InventoryInStorePickupApi/Model/SearchResult/Area/HandleSearchTerm.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+use Magento\Framework\DataObject;
+use Magento\Framework\DataObjectFactory;
+use Magento\InventoryInStorePickupApi\Model\SearchResult\Area\SearchTerm\HandlerInterface;
+use Magento\InventoryInStorePickupApi\Model\SearchResult\StrategyInterface;
+
+/**
+ * Extract address data from Search Term.
+ */
+class HandleSearchTerm
```
Comment (sidolov) [inline InventoryInStorePickupApi/Model/SearchResult/Area/HandleSearchTerm.php]: I think we should rename class with more appropriate name.

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/Parser.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+declare(strict_types=1);
+
+namespace Magento\InventoryInStorePickup\Model\SearchRequest\Area\SearchTerm;
+
+/**
+ * Parse Search Term.
+ */
+class Parser
+{
+    private const DELIMITER = ':';
```
Comment (sidolov) [inline InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/Parser.php]: Please, consider making delimiter configurable.

**Review by sidolov — CHANGES_REQUESTED** [review]

File: InventoryInStorePickupApi/Model/SearchResult/Area/HandleSearchTerm.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+{
+    /**
+     * @var array
```
Comment (novikor) [inline InventoryInStorePickupApi/Model/SearchResult/Area/HandleSearchTerm.php]: `Datatype[]`?

File: InventoryInStorePickupApi/Model/SearchResult/Area/HandleSearchTerm.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    private function validate(array $handlers): void
+    {
+        foreach ($handlers as $handler) {
+            if (!$handler instanceof HandlerInterface) {
+                throw new \InvalidArgumentException(
+                    sprintf(
+                        'Handler must implement %s.' .
+                        '%s has been received instead.',
+                        StrategyInterface::class,
```
Comment (novikor) [inline InventoryInStorePickupApi/Model/SearchResult/Area/HandleSearchTerm.php]: Are you sure `StrategyInterface` is requested?

File: InventoryInStorePickup/Model/SearchRequest/Builder/AreaBuilder.php [hunk 8]
```diff
@@ -16,10 +16,7 @@
 class AreaBuilder
```
Comment (novikor) [inline InventoryInStorePickup/Model/SearchRequest/Builder/AreaBuilder.php:16]: It seems to be a part of API.
Probably it should be moved to API or at least be marked as `@api?`
See `\Magento\Framework\Api\AbstractSimpleObjectBuilder`, `\Magento\Framework\Api\SearchCriteria`, '\Magento\Framework\Api\SearchCriteriaBuilder'.

I believe it should be discussed.

File: InventoryInStorePickup/Model/SearchRequest/Area.php [hunk 9]
```diff
@@ -22,42 +22,18 @@ class Area implements AreaInterface
...
     /**
      * @param int $radius
-     * @param string $country
-     * @param string|null $region
-     * @param string|null $city
-     * @param string|null $postcode
+     * @param string $searchTerm
      */
     public function __construct(
         int $radius,
```
Comment (novikor) [inline InventoryInStorePickup/Model/SearchRequest/Area.php:32]: According to the `AreaBuilder`, any of the parameters can be empty, - so we will have an Exception in this case.

Either validation or nullable type hints should be provided.

**Review by novikor — CHANGES_REQUESTED** [review]

File: InventoryInStorePickup/Model/SearchRequest/Builder/AreaBuilder.php (hunk 8 above)
Comment (swnsma) [reply InventoryInStorePickup/Model/SearchRequest/Builder/AreaBuilder.php:16]: The API only is \Magento\InventoryInStorePickupApi\Model\SearchRequestBuilderInterface
All other builders - implementation.

File: InventoryInStorePickup/Model/SearchRequest/Area.php (hunk 9 above)
Comment (swnsma) [reply InventoryInStorePickup/Model/SearchRequest/Area.php:32]: Could you share a link?

**Comment by novikor** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento run Functional Tests CE,Functional Tests EE,Functional Tests B2B

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CityParser.php [hunk 10]
```diff
@@ -36,9 +38,26 @@ public function __construct(DelimiterParser $parser)
...
     }
+
+    /**
+     * Parse and return search query from search term.
+     *
+     * @param string $searchTerm
+     *
+     * @return string
+     */
+    public function getSearchQuery(string $searchTerm): string
```
Comment (novikor) [inline InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CityParser.php]: Are you sure this method should be public?

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CountryParser.php [hunk 11]
```diff
@@ -22,26 +23,44 @@ class CountryParser implements ParserInterface
...
+    }
+
+    /**
+     * Get country from Search Term.
+     *
+     * @param string $searchTerm
+     *
+     * @return string|null
+     */
+    public function getCountry(string $searchTerm): ?string
```
Comment (novikor) [inline InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CountryParser.php]: Are you sure this method should be public?

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php [hunk 12]
```diff
@@ -24,30 +25,47 @@ class PostcodeParser implements ParserInterface
...
     }
+
+    /**
+     * Parse and return search query from search term.
+     *
+     * @param string $searchTerm
+     *
+     * @return string
+     */
+    public function getSearchQuery(string $searchTerm): string
```
Comment (novikor) [inline InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php]: This method is duplicated 3 times. Something went wrong.

**Review by novikor — CHANGES_REQUESTED** [review]
`InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CityParser.php`
`InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CountryParser.php`
`InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php`

Well, except duplicated code (`getSearchQuery`, `getCountry`, `getSearchQuery`) it looks strange when a class has any public methods while having `execute`, - it`s not MSI-style service anymore.

You should either rename `execute` either move the public methods.

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CityParser.php (hunk 10 above)
Comment (swnsma) [reply InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CityParser.php]: Public -> private.

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CountryParser.php (hunk 11 above)
Comment (swnsma) [reply InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/CountryParser.php]: Public -> private.

File: InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php (hunk 12 above)
Comment (swnsma) [reply InventoryInStorePickup/Model/SearchRequest/Area/SearchTerm/PostcodeParser.php]: 2 times*
It was agreed to do not use "helper" classes.

**Review by novikor — APPROVED** [review]

File: InventoryInStorePickupApi/Model/SearchRequest/Area/Pipeline.php [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * @param string $searchTerm
+     * @return DataObject
+     */
+    public function execute(string $searchTerm) : DataObject
+    {
+        $result = $this->dataObjectFactory->create();
+        foreach ($this->parsers as $parser) {
+            $parser->execute($searchTerm, $result);
```
Comment (ishakhsuvarov) [inline InventoryInStorePickupApi/Model/SearchRequest/Area/Pipeline.php:70]: Is this really a pipeline implementation? as I would expect result from each iteration to be sent over to the next one without modification

Comment (novikor) [reply InventoryInStorePickupApi/Model/SearchRequest/Area/Pipeline.php:70]: @ishakhsuvarov , - it`s not a real pipeline you used to face with: the result is being modified indeed.
However we can not call it a chain, - the sequence is important in this case.

**Comment by sidolov** [comment]
@magento run Magento Health Index

**Review by sidolov — APPROVED** [review]

**Pull request merged**